### ISSUE TITLE
Make i3-dmenu-desktop startup faster

### ISSUE THREAD

**Issue by i3bot** [i3/i3#949]
[Originally reported by gereon.kremer@…]
(i3-dmenu-desktop takes some 100ms to startup. This is not really slow, but in this case already annoying: When you start typing down your application directly after launching i3-dmenu-desktop, you can very well be faster than the launcher and the first few keystrokes are lost.
This is on a system with <200 .desktop files in /usr/share/applications/ on a SSD... so I guess it might be even worse on other systems.

At the same time, I see a relatively easy approach to get a major speedup: As far as I understand the script (I never used perl...), it parses all desktop-files everytime.
Couldn't we just do it once and cache the result somewhere in /tmp/?

**Comment by kilbith** [comment]
@stapelberg Is it possible to know why `j4-dmenu` has not been included in i3, please ? The author claims and proves it's approx. 25x faster than `i3-dmenu`.

**Comment by stapelberg** [comment]
@kilbith The author of j4-dmenu-desktop has not made any efforts (AFAIR) to get it merged into i3.

Given that j4-dmenu-desktop is independent of the window manager, I’m not sure bundling it with the i3 source code is a good idea.

What is the benefit for you of including j4-dmenu-desktop in i3?

One way forward that I see is deleting i3-dmenu-desktop and making j4-dmenu-desktop a dependency of i3 (provided j4-dmenu-desktop is a drop-in replacement).

**Comment by kilbith** [comment]
@stapelberg `j4-dmenu` seems slightly snappier than `i3-dmenu` for me. But maybe it's just a placebo effect... The best of it is that it's agnostic from WMs.

Thanks fo your answer.

**Comment by TT44bb** [comment]
On my slow netbook, i3-dmenu-desktop takes between 1s and 1,2s to launch (measured using: time i3-dmenu-desktop --dmenu=error). This is not acceptable, it makes it useless.

time j4-dmenu-desktop --dmenu=error consistently takes less than 0.050s, usually rather 0.030s.

Bundling j4-dmenu with the code is not what I'm suggesting, but instead that commented out line in the default config file about i3-dmenu should be replaced with the equivalent for j4-dmenu and j4-dmenu should be a dependency or recommended.

**Comment by Airblader** [comment]
>  j4-dmenu should be a dependency or recommended.

That's not up to us, that's up to the package maintainer.

> but instead that commented out line in the default config file about i3-dmenu should be replaced with the equivalent for j4-dmenu

I'd prefer to suggest i3 builtins over external tools for anything where we have an i3 builtin. External tools are not under our control.

**Comment by TT44bb** [comment]
I doubt the package maintainer will do this unless you suggest it.

> I'd prefer to suggest i3 builtins over external tools for anything where we have an i3 builtin. External tools are not under our control.
Sure. If the i3 built ins work. If they don't work, they shouldn't be built ins.

**Comment by Airblader** [comment]
> I doubt the package maintainer will do this unless you suggest it.

We have no »this is what we recommend« list anywhere. The Debian package recommends dunst which we never mention anywhere.

> If the i3 built ins work. If they don't work, they shouldn't be built ins.

If they don't work, file a bug. I haven't seen anything that states that it doesn't work.

A dependency forced by us means i3 absolutely needs that tool. That is simply not true for j4. If a package maintainer wants to provide it that is their decision. To have j4 as a hard dependency would require it to become part of the project. As @stapelberg pointed out, nobody has made any efforts in that direction.

**Comment by Airblader** [comment]
On a side note: I'd personally actually prefer to split i3 more. In my opinion i3-dmenu-desktop shouldn't be part of core i3 (due to the suggested dependency to dmenu), neither should be i3bar and i3-input. Just my opinion, though.

**Comment by TT44bb** [comment]
I don't really know about splitting it up more or making everything one project. I'm more of a user who would have like something like i3-dmenu-desktop without having to search for j4-dmenu. A 1s start up for a program like this doesn't count as "works" if you ask me, and requiring a fast computer and an ssd for it is just laughable.

**Comment by Airblader** [comment]
Nobody said we are against making it faster. @stapelberg even said it'd be possible to merge the j4 version. But at least for the past three years nobody has made any contribution in that direction.

You explicitly mentioned making j4 a dependency and I'm simply explaining why that's the wrong approach. Anyone who wants to make i3-dmenu-desktop is more than welcome to do so and provide a pull request.

**Comment by kilbith** [comment]
You're a maintainer, it's more your responsability to do that than a random Joe doing a PR, eventually. Don't let your crap of `i3-dmenu` there until someone dares to move a finder by chance, please.

**Comment by Airblader** [comment]
i3 is a community-driven project. If the community is not interested in participating, we don't feel obiliged to fill the gap.

**Comment by TT44bb** [comment]
I don't really know what I could do to help. It's more of an organisational issue than anything else, after all working code exists. It just the point of making people aware of it. I wouldn't know j4-dmenu existed if I hadn't tried to file a bug report for the included one being so slow.

What about, for now, just adding a comment to the default config file saying that i3-dmenu is slow and that j4-dmenu exists and is faster?

**Comment by Airblader** [comment]
> What about, for now, just adding a comment to the default config file saying that i3-dmenu is slow and that j4-dmenu exists and is faster?

Any solution that contains »for now« considers itself a workaround and those have a nasty habit of sticking around. Also we want to avoid bloating the config because it decreases readability for newcomers overall.

FYI, there are many more (and in my personal opinion considerably better) options for launchers such as rofi, lighthouse, …

**Comment by TT44bb** [comment]
> We have a clear vision of what the proper solution is, why not work on that?

We do? Not so sure anymore.

**Comment by stapelberg** [comment]
First of all, I’d like to ask all participants in this discussion to stay civil. Labeling our work as “crap” certainly doesn’t help motivating me to spend time on this :(.

j4-dmenu-desktop is not actually i3-specific since v2.7, so merging it into the i3 codebase might not actually be the best solution.

I’m not sure if j4-dmenu-desktop is a drop-in solution for i3-dmenu-desktop (i.e. flag-for-flag compatible, and with the same default behavior). If it is, I’m willing to deprecate i3-dmenu-desktop and add a recommendation for j4-dmenu-desktop to the i3 package instead. Can anyone investigate and confirm please?

**Comment by TT44bb** [comment]
It's not a drop-in solution. --entry-type=type is missing from j4-dmenu and the default behaviour is different too (it's basically like --entry-type=name).

So it seems like we have a course of action: modify j4-dmenu so it becomes a drop-in replacement.

**Comment by Airblader** [comment]
@stapelberg What exactly would the deprecation look like? How is a user informed of it and will this require the 5.0 release?

**Comment by stapelberg** [comment]
We’d make the appropriate changes in the default config file and add notes about that in the release notes/i3-dmenu-desktop itself.

We might want to add a warning/info notice to i3-dmenu-desktop, but given that it’s not executed on a terminal, people will likely not see that.

Eventually, we could turn i3-dmenu-desktop into a symlink to j4-dmenu-desktop.

I don’t think that a 5.0 release will be necessary — we’re not breaking anything, provided that j4-dmenu-desktop is a drop-in replacement (or close enough for the majority of use-cases).

**Comment by Airblader** [comment]
> I don’t think that a 5.0 release will be necessary — we’re not breaking anything, provided that j4-dmenu-desktop is a drop-in replacement (or close enough for the majority of use-cases).

Only if we actually depend on j4-dmenu-desktop, which isn't what I've understood (nor prefer). A user using i3-dmenu-desktop will see breakage if we stop shipping i3-dmenu-desktop and they don't have j4-dmenu-desktop installed.

**Comment by stapelberg** [comment]
I see what you’re saying. Let’s expand a bit more:

## Migration strategy from the i3 point of view

In the next major version, we would change our configuration file to use j4-dmenu-desktop.

In the following major version, we could add a check to our config handling which verifies whether you have any bindings that match the string “i3-dmenu-desktop” and show a deprecation notice, much like the v3→v4 config migration was handled.

We could, in addition (possibly one major version later), have i3-dmenu-desktop print a deprecation notice in its prompt — that is bound to be seen by every user.

Then, a major version later, we could delete i3-dmenu-desktop. People who are attached to it for whatever reason can get it from the git history and it’s easy to drop it into a system.

## Migration strategy from the package maintainer point of view

Depending on the Linux distribution in question, I think it would be reasonable to either:

1. For Debian and derivatives: Add a Recommends on j4-dmenu-desktop so that people who have Recommends enabled (the default) get the package and things will just keep working for them.
2. For distributions without the Recommends relation, we could have an alternative dependency that defaults to j4-dmenu-desktop (so that for new installations, the right thing happens) or i3-dmenu-desktop (so that existing installations are unchanged).

Point ② implies creating a new package just for i3-dmenu-desktop, so some package maintainers might not want to do that. I think in that case they should just add a dependency on j4-dmenu-desktop. Ideally, they already have the concept of an “i3” package (i3 plus common tools) vs. an “i3-wm” package (the window manager itself).

**Comment by Airblader** [comment]
> In the next major version […]

You mentioned several »next major« versions. Do you actually mean 5.x, 6.x, …? Or do you consider 5.13, 5.14, … major versions here?

Personally I'm against such immense efforts for something semantic versioning can handle perfectly fine, but given that I won't be working on this myself anyway I'm not going to object if somebody wants to go through the trouble to do this. :-)

**Comment by stapelberg** [comment]
I meant 4.13, 4.14, etc.

> I'd really only like to avoid a hard dependency on j4-dmenu-desktop. Plenty of users don't use it and we shouldn't force them to install it. We don't even depend on i3status, although the typical first installation by a user will display an error in i3bar because of that.

This is distribution-specific. In Debian, the i3 package does pull in i3status, and j4-dmenu-desktop should be treated the same way as i3status, I’d say.

**Comment by Airblader** [comment]
I consider it a mistake to force a tool on the user they don't need to use. But I don't use Debian and my distro doesn't force it. In the end it's a package maintainer issue. As long as we don't communicate that we depend on it and leave that up to the distros it's fine with me.

**Comment by TT44bb** [comment]
@Airblader  That's true, but it's also important to tell the user that this tool *exists*, otherwise they can't decide whether they want to use it.
I'm pretty sure I wouldn't use i3 today if i3status hadn't been pulled in automatically by apt, because it would have seemed to be lacking to many important features. On the other hand, with i3status preinstalled, the out-of-the-box experience was good enough that I thought it was worth it to look more into it and find out about stuff like j4-dmenu so I could further improve my setup.

I don't know, but I guess i3status uses on the order of kilobytes of hard disk space. And when someone doesn't want to use it, they can just disable it from the config file. Of course that's not ideal, but I think considering the existing alternatives it's best. The average user wouldn't care about kilobytes of hard disk and someone who is *so* hard pressed for space would probably be running a completely custom built setup anyway.

**Comment by Airblader** [comment]
I agree that i3status is small enough to not hurt in a way where disk space is any kind of concern. It's more of a modularization problem. i3status is, in principle, a separate module (which is a good architecture), but there's a rather unfortunate dependency in the default config which breaks this separation. Personally, my ideal target architecture would be to have libi3 as a separate library and then have i3, i3bar and i3status as separate modules depending only on libi3. Even something like i3-msg could probably be refactored into a module.

And regarding your first point: yes, a user needs to know about their options. The question is just how that can be achieved. We could also list all kinds of options in the default config, but that is prone to running out of date and at some point too many options are more confusing than helpful. It's a fine line to walk on. In general given that i3 is aimed at power users I tend to take the stance that I expect i3 users to do some reading and research. If our target audience were Unix or even computer beginners, that'd be different; but they're not.

**Comment by stapelberg** [comment]
> Personally, my ideal target architecture would be to have libi3 as a separate library and then have i3, i3bar and i3status as separate modules depending only on libi3

While I do see the motivation behind your suggestion and agree with it in principle, I think that currently, the downsides (having to deal with shared libraries) far outweigh the gains (slight reduction of duplicated code).

**Comment by Airblader** [comment]
Yeah, I don't have any plans to go in that direction (otherwise you'd know about them :-)) because it's a fair amount of work both on our part and on the part of all distros.

**Comment by hholst80** [comment]
A quick solution to this as I don't see any PRs or commits in this issue yet: Launch the application and accept keyboard input directly. Populate the .desktop and executables via a background thread. That should solve all problems. If you would be open to rewriting i3-dmenu-desktop in Python I could contribute code. ;-)

Side note: The delay right now from pressing "launch menu application" to the point where it accepts input, on the first startup, is over 300ms on my machine. My machine has 56 cores;  64 GiB ram and a $1000 dollar SSD drive.

**Comment by fridim** [comment]
> A quick solution to this as I don't see any PRs or commits in this issue yet: Launch the application and accept keyboard input directly. Populate the .desktop and executables via a background thread. That should solve all problems.

I really like this solution.

BTW, a quick profiling:
```                                                                                                                    
[fridim@lolo profiling]$ perl -d:NYTProf /usr/bin/i3-dmenu-desktop --dmenu=cat                                                                                                                 
[{"success":true}]                                                                                                                                                                            
[fridim@lolo profiling]$ nytprofhtml --open                                                                                                                                                   
Reading nytprof.out                                                                                                                                                                           
Processing nytprof.out data                                                                                                                                                                   
Writing line reports to nytprof directory                                                                                                                                                     
 100% ...                                                                                                                                                                                     
Extracting subroutine call data ...                                                                                                                                                           
Extracting subroutine links                                                                                                                                                                   
Generating subroutine stack flame graph ...  
```
results:

```
Top 15 Subroutines Calls 	P 	F 	Exclusive
Time 	Inclusive
Time 	Subroutine
11656	3	1	55.5ms	156ms	          File::Basename::basename
11656	1	1	47.3ms	54.8ms	          File::Basename::fileparse
1	1	1	41.6ms	41.6ms	              IPC::Open3::CORE:read (opcode)
11656	1	1	31.0ms	45.9ms	          File::Basename::_strip_trailing_sep
11656	1	1	14.9ms	14.9ms	          File::Basename::CORE:subst (opcode)
15751	3	1	10.5ms	10.5ms	                    main::CORE:match (opcode)
1	1	1	8.80ms	10.0ms	                    main::BEGIN@13
11666	2	1	7.44ms	7.44ms	          File::Basename::CORE:match (opcode)
1	1	1	6.65ms	6.69ms	Pod::Simple::LinkSection::BEGIN@9
1	1	1	5.73ms	14.8ms	               Pod::Text::BEGIN@35
1	1	1	5.73ms	6.35ms	                    main::BEGIN@16
473	1	1	4.47ms	4.66ms	            Encode::utf8::decode_xs (xsub)
1	1	1	4.34ms	10.5ms	                    main::BEGIN@10
1	1	1	4.31ms	10.9ms	                    main::BEGIN@15
1	1	1	3.67ms	3.88ms	            Data::Dumper::BEGIN@22
```
11656 is a lot of calls. `basename` is using a lot of cpu time. There are some complicated regex there...

For testing purpose, I patched i3-dmenu-desktop to add caching on `%apps` and `%choices` using [Storable](http://perldoc.perl.org/Storable.html): https://gist.github.com/fridim/016dfc9d0ae03f04729342323825227f

```
[fridim@lolo tmp]$ time ./i3-dmenu-desktop --store=i3-dmenu-desktop.db --dmenu=cat                                                                                                            
[{"success":true}]                                                                                                                                                                            
                                                                                                                                                                                              
real    0m0.176s                                                                                                                                                                              
user    0m0.160s                                                                                                                                                                              
sys     0m0.010s                                                                                                                                                                              
[fridim@lolo tmp]$ time ./i3-dmenu-desktop --load=i3-dmenu-desktop.db --dmenu=cat                                                                                                             
[{"success":true}]                                                                                                                                                                            
                                                                                                                                                                                              
real    0m0.076s                                                                                                                                                                              
user    0m0.067s                                                                                                                                                                              
sys     0m0.007s
```

In addition to the 'allow input as soon as possible' solution, i think it would be convenient to have an option like `--cache` that would use a previous calculated list of .desktop (so it can starts quickly), and then update the cache in background when exiting. The user will have to run i3-dmenu-desktop twice after installing a new application though.

**Comment by stapelberg** [comment]
Thanks for looking into this.

I have to say, I’m not sure what the benefit of the proposed `--cache` option is: unless it’s active by default (without side-effects such as having to start twice after installing new applications, though!), the effort for the end-user is very similar to just selecting j4-dmenu-desktop in their config file.

**Comment by fridim** [comment]
Indeed, i was focusing on my use-case. i3-dmenu-desktop is used by default on qubes-os (when i3 is chosen as a WM). I just didn't want to compile and install in dom0 some code not audited.

**Comment by EgbertW** [comment]
For me, i3-dmenu-desktop takes 0.694s to appear, j4-dmenu-desktop takes 0.007s.

I'm consequently missing several characters because I start typing too soon. Thanks for the j4-dmenu-desktop suggestion, I have switched. Would be cool though to have it supersede i3-dmenu-desktop completely - it's not like it's a big dependency - entire repository on GitHub takes 1.2MiB.

Anyway, glad I found this issue because I have a solution now!

**Comment by tboby** [comment]
Wow, as someone who isn't a linux power-user, I'd been assuming the fault lay with my laptop's hardware. 

Replacing i3-dmenu-desktop with j4-dmenu-desktop took the first run menu responsiveness from several seconds to instant.

Should the default config file perhaps mention j4-dmenu-desktop, to signpost users to the best default experience?

**Comment by Airblader** [comment]
As mentioned somewhere above, I don't much like pointing to alternatives that are not in our control. It may not be available on the user's system, it may no longer exist at some point, there may be other alternatives etc.

**Comment by tboby** [comment]
Well, I'd suggest there's a difference between expecting users to find an alternative to `dmenu`, and expecting users to know about an alternative to a part of the core i3 library :)

It just seems a shame to require users to do nitty gritty research (looking through the issue tracker, not just the config/user guide/FAQ) to fix a known defect with a simple workaround. A defect caused by a recommendation in the config!

**Comment by megatux** [comment]
Hi. I'll try to explain my issue.
I'm using Regolith desktop, so AFAIK calling the menu is invoking rofi. I have 359 .desktop files and a HDD.  A cold startup takes like **10 seconds** to show the menu!  ...but my main issue is not SPEED. I can wait. My main issue is that the rest of the WM is **stall** while it _process_ the menu. I'm unable to switch workspaces, use other windows, etc.  
Is there a way to do the call in an _async_ way?
For the record, the rest of the DEs also  takes several seconds to build the apps menu.
Thanks a lot!
